flocker/control/_model.py

```diff
@@ -451,7 +482,10 @@ def __new__(cls, hostname=None, **kwargs):
         return PClass.__new__(cls, **kwargs)
 
     uuid = field(type=UUID, mandatory=True)
-    applications = pset_field(Application)
+    applications = pmap_field(
+        unicode, Application, invariant=_keys_match("name"),
+        factory=lambda x: _turn_iterable_to_mapping_from_attribute('name', x)
```

Comment:
Just remind me of the motivation for this. 
Are just trying to avoid changing existing code that supplies Application sets rather than mappings.
It seems confusing to be able to supply a list of applications here:

```
In [112]: Node(uuid=uuid4(), applications=[Application(name=u"database1", image=DockerImage(repository="clusterhq/foodb", tag="latest"))])
Out[112]: Node(applications=UnicodeApplicationPMap({u'database1': Application(memory_limit=None, name=u'database1', links=LinkPSet([]), environment=pmap({}), command_line=None, image=DockerImage(tag='latest', repository='clusterhq/foodb'), restart_policy=RestartNever(), volume=None, running=True, ports=PortPSet([]), cpu_shares=None)}), manifestations=UnicodeManifestationPMap({}), uuid=UUID('4701d23b-de5e-4e85-9e57-431345cecd29'))

In [113]: Node(uuid=uuid4(), applications=[Application(name="database1", image=DockerImage(repository="clusterhq/foodb", tag="latest"))])
---------------------------------------------------------------------------
CheckedKeyTypeError                       Traceback (most recent call last)
<ipython-input-113-1f36d68ca295> in <module>()
----> 1 Node(uuid=uuid4(), applications=[Application(name="database1", image=DockerImage(repository="clusterhq/foodb", tag="latest"))])

/home/richard/projects/HybridLogic/flocker/flocker/control/_model.py in __new__(cls, hostname, **kwargs)
    480                  "a bug.", DeprecationWarning, stacklevel=2)
    481             kwargs["uuid"] = ip_to_uuid(hostname)
--> 482         return PClass.__new__(cls, **kwargs)
    483 
    484     uuid = field(type=UUID, mandatory=True)

/home/richard/.virtualenvs/4417/lib/python2.7/site-packages/pyrsistent/_pclass.pyc in __new__(cls, **kwargs)
     50         for name, field in cls._pclass_fields.items():
     51             if name in kwargs:
---> 52                 value = field.factory(kwargs[name])
     53                 _check_and_set_attr(cls, field, name, value, result, invariant_errors)
     54                 del kwargs[name]

/home/richard/projects/HybridLogic/flocker/flocker/control/_model.py in <lambda>(x)
    136 
    137     if input_factory:
--> 138         factory = lambda x: mapping_factory(input_factory(x))
    139     else:
    140         factory = mapping_factory

/home/richard/.virtualenvs/4417/lib/python2.7/site-packages/pyrsistent/_checked_types.pyc in __new__(cls, initial, size)
    409         evolver = CheckedPMap.Evolver(cls, pmap())
    410         for k, v in initial.items():
--> 411             evolver.set(k, v)
    412 
    413         return evolver.persistent()

/home/richard/.virtualenvs/4417/lib/python2.7/site-packages/pyrsistent/_checked_types.pyc in set(self, key, value)
    457 
    458         def set(self, key, value):
--> 459             _check_types([key], self._destination_class._checked_key_types, self._destination_class, CheckedKeyTypeError)
    460             _check_types([value], self._destination_class._checked_value_types, self._destination_class)
    461             self._invariant_errors.extend(data for valid, data in (invariant(key, value)

/home/richard/.virtualenvs/4417/lib/python2.7/site-packages/pyrsistent/_checked_types.pyc in _check_types(it, expected_types, source_class, exception_type)
    171                     expected_types=tuple(get_type(et).__name__ for et in expected_types),
    172                     actual_type=actual_type.__name__)
--> 173                 raise exception_type(source_class, expected_types, actual_type, e, msg)
    174 
    175 

CheckedKeyTypeError: Type UnicodeApplicationPMap can only be used with ('unicode',), not str


```

Also note that Application seems to allow a name to be str where the pmap_field only accepts unicode.